core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -532,6 +534,23 @@ private void innerIndex(Index index) throws IOException {
         }
     }
 
+    /**
+     * Asserts that the doc in the index operation really doesn't exist
+     */
+    private boolean assertDocDoesNotExist(Index index) throws IOException {
+        final VersionValue versionValue = versionMap.getUnderLock(index.uid());
+        if (versionValue != null && versionValue.delete() == false) {
+            throw new AssertionError("doc [" + index.type() + "][" + index.id() + "] exists in version map (version " + versionValue + ")");
+        }
+        try (final Searcher searcher = acquireSearcher("assert doc doesn't exist")) {
+            final long docsWithId = searcher.searcher().count(new TermQuery(index.uid()));
```

Comment:
yeah this looks better but I think we can still have a doc in the index when we haven't refreshed AND the doc is deleted? I think we should not check the searcher if `versionValue.delete() == true`?